utils.py

```diff
@@ -253,3 +315,222 @@ def availability_to_dict(availability_list):
+        understaffed_key, understaffed_template = msgs["understaffed"]
+        overstaffed_key, overstaffed_template = msgs["overstaffed"]
+        num_employees = results[:, s].sum()
+        if num_employees < params.shift_min:
+            errors[understaffed_key].append(understaffed_template.format(day=day))
+        elif num_employees > params.shift_max:
+            errors[overstaffed_key].append(overstaffed_template.format(day=day))
+    return errors
+
+
+def _validate_requires_manager(
+    params: ModelParams,
+    results: np.ndarray,
+    employees: list[str],
+    errors: defaultdict[str, list[str]],
+    msgs: dict[str, tuple[str, str]],
+) -> defaultdict[str, list[str]]:
+    """Validates the number of managers per shift for the solution and updates
+    the `errors` dictionary with any errors found. Requires the `msgs` dict
+    to have the key `'manager_issue'`."""
+    key, template = msgs["manager_issue"]
+    employee_arr = np.asarray(
+        [employees.index(e) for e in employees if e[-3:] == "Mgr"]
+    )
+    managers_per_shift = results[employee_arr].sum(axis=0)
+    for shift, num_managers in enumerate(managers_per_shift):
+        if num_managers == 0:
+            errors[key].append(template.format(day=params.shift_labels[shift]))
+    return errors
+
+
+def _validate_isolated_days_off(
+    params: ModelParams,
+    results: np.ndarray,
+    employees: list[str],
+    errors: defaultdict[str, list[str]],
+    msgs: dict[str, tuple[str, str]],
+) -> defaultdict[str, list[str]]:
+    """Validates the number of managers per shift for the solution and updates
+    the `errors` dictionary with any errors found. Requires the `msgs` dict
+    to have the key `'isolated'`."""
+    key, template = msgs["isolated"]
+    isolated_pattern = np.array([1, 0, 1])
+    for e, employee in enumerate(employees):
+        shift_triples = [results[e, i : i + 3] for i in range(results.shape[1] - 2)]
+        for s, shift_set in enumerate(shift_triples):
+            if np.equal(shift_set, isolated_pattern).all():
+                day = params.shift_labels[s + 1]
+                errors[key].append(template.format(employee=employee, day=day))
+    return errors
+
+
+def _validate_max_consecutive_shifts(
+    params: ModelParams,
+    results: np.ndarray,
+    employees: list[str],
+    errors: defaultdict[str, list[str]],
+    msgs: dict[str, tuple[str, str]],
+) -> defaultdict[str, list[str]]:
+    """Validates the max number of consecutive shifts for the solution and updates
+    the `errors` dictionary with any errors found. Requires the `msgs` dict
+    to have the key `'too_many_consecutive'`."""
+    key, template = msgs["too_many_consecutive"]
+    for e, employee in enumerate(employees):
+        for shift, shift_arr in enumerate(
+            [
+                results[e, i : i + params.max_consecutive_shifts]
+                for i in range(results.shape[1] - params.max_consecutive_shifts)
+            ]
```

Comment:
Super minor but I find this a bit hard to read so it might be nice to pull it out into a variable

Reply:
Pulled this out into `consecutive_shift_arrays`